File: Algo/Leetcode/114FlattenBinaryTreetoLinkedList.cpp
//1. Flatten both left and right. Move left to the right.
class Solution {
public:
    void flatten(TreeNode* root) {
        if(!root) return;
        if(!root->left && !root->right) return;
        
        flatten(root->left);
        flatten(root->right);
        
        if(!root->left) return;
        
        TreeNode* Lend = root->left;
        while(Lend->right){
            Lend = Lend->right;
        }
        Lend -> right = root->right;
        root -> right = root ->left;
        root ->left = nullptr;
    }
};

// same dfs
class Solution {
public:
    void flatten(TreeNode* root) {
        if(!root) return;
        dfs(root);
    }

    void dfs(TreeNode* root){
        if(!root) return;
        dfs(root->left);
        dfs(root->right);
        if(root->left){
            TreeNode* tmp = root->left;
            while(tmp->right)
                tmp = tmp->right;
            tmp->right = root->right;
            root->right = root->left;
            root->left = NULL;
        }
    }
};

File: Algo/Leetcode/114FlattenBinaryTreetoLinkedList.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: None Do not return anything, modify root in-place instead.
        """
        if root is None:
            return
        if root.left is None and root.right is None:
            return root

        # inplace
        self.flatten(root.left)
        self.flatten(root.right)

        if root.left is None:
            return

        l_tail = root.left
        while l_tail.right:
            l_tail = l_tail.right
        l_tail.right = root.right

        # reset
        root.right = root.left
        root.left = None
